pkg/adapters/rules/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,284 @@
+	// and is written by the cache event handlers (internally).
+	// Given that we expect a single reader only, we use a regular sync.Mutex rather than a sync.RWMutex.
+	mutex sync.Mutex
+}
+
+// New creates and starts a new Kubernetes Rules adapter.
+func New(config Config) (*Adapter, error) {
+	logger.Info("Adapter starting..." )
+	namespace := config.Namespace.String()
+	// If no namespace is specified, fallback to default namespace
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	// Create the kubernetes client for the rules resource
+	client, err := createRulesClient(config)
+	if err != nil {
+		return nil, err
+	}
+	workqueue := kubepkg.NewWorkqueue()
+
+	adapter := &Adapter{
+		client:    client,
+		workqueue: workqueue,
+		rules:     make(map[string]*a8api.Rule),
+		namespace: namespace,
+		revision: 0,
+	}
+	adapter.rulesCache, adapter.rulesController = cache.NewInformer(
+		cache.NewListWatchFromClient(client, fmt.Sprintf("%ss", strings.Replace(ResourceName, "-", "", -1)), namespace, nil),
+		&RoutingRule{},
+		RulesCacheResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    workqueue.EnqueueingAddFunc(adapter.addRule),
+			UpdateFunc: workqueue.EnqueueingUpdateFunc(adapter.updateRules),
+			DeleteFunc: workqueue.EnqueueingDeleteFunc(adapter.deleteRules),
+		},
+	)
+	logger.Info("Calling Adapter start" )
+	adapter.workqueue.Start()
+
+	return adapter, adapter.Start()
+}
+
+// Start synchronizing the Kubernetes adapter.
+func (a *Adapter) Start() error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	if a.stopChan != nil {
+		err := fmt.Errorf("kubernetes adapter already started")
+
+		logger.WithError(err).Errorf("Failed starting Kubernetes adapter")
+		return err
+	}
+	a.stopChan = make(chan struct{})
+
+	go a.rulesController.Run(a.stopChan)
+	return nil
+}
+
+func createRulesClient(config Config) (*rest.RESTClient, error) {
+	var kubeConfig *rest.Config
+	if config.URL != "" {
+		kubeConfig = &rest.Config{
+			Host:        config.URL,
+			BearerToken: config.Token,
+		}
+	} else {
+		var err error
+		kubeConfig, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed loading Kubernetes credentials from service account")
+		}
+	}
+
+	groupversion := unversioned.GroupVersion{
+		Group:   ResourceGroupName,
+		Version: ResourceVersion,
+	}
+
+	kubeConfig.GroupVersion = &groupversion
+	kubeConfig.APIPath = "/apis"
+	kubeConfig.ContentType = runtime.ContentTypeJSON
+	kubeConfig.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				groupversion,
+				&RoutingRule{},
+				&RoutingRuleList{},
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	client, err := rest.RESTClientFor(kubeConfig)
+	if err != nil {
+		logger.Printf("Failed to create the in-cluster client. %s\n", err)
```

Comment:
We use `logger.WithError(err).Errorf(...)` for logging errors.

Also note that the log message is not accurate - the client is not necessarily in-cluster...